### ISSUE TITLE
[Feature request] Add before consent / after consent gestion

### ISSUE THREAD

**Issue by SW-Vincent** [matomo-org/tag-manager#761]
While managing consent is becoming a chalenge over time, it would be nice to have dedicated features such as : 
 - tagging variables as consent variables, these consent variables will then be declared within the tags as the tested consent variable for this tag ; 
 - tagging trigger as consent update trigger, this trigger will then be declared within the tags as the the  ;
 - within tag template, declaring what needs to be done before consent and what needs to be done after consent* ;
Eventually, make the third point an option (ex : throught a checkbox)

*For the last point, it basically means that each tag will actually be x tags in one with one trigger and one trigger group : 
 - First tag (before consent) will trigger when the defined trigger is matched.
 - Second tag -eventually more- (after consent) will trigger either when the default trigger is matched and the consent variable is "true" OR after the default trigger has been triggered if the consent trigger is matched and the consent variable is "true". It will have to be triggered once per default trigger, weither the consent trigger has been matched before or after them.

Note that for such a thing to be reliable for SPA, we'll need some data layer "reset" mechanic.

**Comment by snake14** [comment]
Hi @SW-Vincent . Thank you for taking the time to submit this enhancement request. I'm marking it to be reviewed by our Product team. Any additional information or clarification you can provide would be appreciated.

**Comment by SW-Vincent** [comment]
Hi,

Here is an exemple of how it might look on a custom HTML Tag :

Below the usual custom HTML tag definition, a CTA would allow to add a new section within the same tag : 
![image](https://github.com/matomo-org/tag-manager/assets/118832007/3bd3f13a-086f-44b3-9879-116df1281839)
Maybe this CTA should be after the trigger section ?

Anyways, the rest is pretty mutch the same that adding any custom HTML tag with slight differences as it requires :
 - the HTML content ;
 - consent triggers (although the consent will also be tested at default triggers), it would make sense to have advanced settings (such as "once per page", etc.) for consent triggers too
- The consent variable(s) and a matching value (or regex ?) 
![image](https://github.com/matomo-org/tag-manager/assets/118832007/a255ff47-9396-48e3-9a8e-4ea11a593527)

Note that it would make sense to be able to add as much consent sections as you want.

If you need some user case for a Matomo Tag, here is what would be wonderful : 
 - creating a Matomo Pageview tag, with a default parameter "require consent for heatmaps and session recordings", "require consent for e-commerce", "require consent for Youtube videos tracking" (note that such parameters don't exist at the moment) ;
 - creating a tag to reactivate each consent, with a consent update trigger and a dedicated consent variable for each ;

In a nutshell, it would be pretty much like Google Consent Mode but better, as you can chose what is sent before and after consent freely and for any tag.

**Comment by Stan-vw** [comment]
Ticket was closed accidentally through an integration. Reopening.

For this ticket, we're looking at implementing the following:
1. Create a dedicated "consent status" variable to hold the consent status
2. Create a dedicated "consent changed" trigger which fires when "consent status" is changed.

This can then be used in a variety of generic ways, such as turning off/on entire containers depending on this status.

Further to this, we are considering:
- if a tag's trigger config could/should then have the option to be configured depending on variable status, such that you could say "this tag triggers when pageview+consent=yes" or such
- to make generic a version of this "variable changed" trigger, which could be used for a myriad of things

However, to keep scope small in an effort to optimise for value added, we keep this out of scope and they can be designed after we have some more experience using the dedicated consent status and triggers.

Keen to read your thoughts on this. Would the in-scope 2 items satisfy the trouble you're bumping into?

**Comment by SW-Vincent** [comment]
Hi Stan,

These features would come in handy !

> if a tag's trigger config could/should then have the option to be configured depending on variable status, such that you could say "this tag triggers when pageview+consent=yes" or such

Do you mean something like the trigger group feature in GTM that would make the tag wait for both pageview trigger and consent trigger to occur ?
It comes with an additionnal concern : what about single page applications ? Note that the GTM page group feature i am comparing this to doesn't handle them.

Additionnally, i am wondering weither Matomo Tag Manager should (or not) be able to read GTM dataLayer after initialization.
Indeed, most Consent Management Platforms will send consent variables and triggers at a GTM format (window.dataLayer variable) and at the moment Matomo Tag Manager can only read what happened before it was initialized.
In my case, when i want to copy GTM dataLayers to Matomo Data Layers, i use something like that (after disabling the default behavior) : 
```
<script>

window.dataLayer = window.dataLayer || [];

function handleGtmDl(data) {
  //if you want to copy some events specifically rather than copying the whole dataLayer variable
  if (data && data.event && data.event === "specificEvent") {  
    _mtm.push(data)
  }
}

//MonkeyPatch the original function with the copy function to handle future events
var originalDataLayerPushTest = window.dataLayer.push;

window.dataLayer.push = function() {
  originalDataLayerPush.apply(window.dataLayer, arguments);
  handleGtmDl(arguments[0]);
};

// Execute the copy function to current dataLayer variable to copy past events
for (var i = 0; i < window.dataLayer.length; i++) {
  handleGtmDl(window.dataLayer[i]);
}

</script>
```

**Comment by Stan-vw** [comment]
Thanks @SW-Vincent . To answer your questions:
1. yes, the idea is to add something like "Only fire this tag if VARIABLE_NAME = XYZ" in the "Configure when the tag should do this" section
2. Can you check if your GTM-sync concern is fixed with https://github.com/matomo-org/tag-manager/pull/879 that was recently created?

**Comment by Stan-vw** [comment]
I've created https://github.com/matomo-org/tag-manager/issues/892 for the specific idea of adding the ability to adjust tag firing logic using variables.

Today I learned that many people & systems pull their consent status straight from the data layer, and users can of course already store that in a custom variable rather than a dedicated named variable. So perhaps having a dedicated consent status variable and a related consent status change trigger aren't as useful.

@SW-Vincent would those still help you, or would it be more helpful to spend time on the other ticket instead?

Btw, as I'm re-reading the original message again I see that we've moved away from the original idea quite a bit. Feel free to explain in more detail what the exact painpoint is, so that we can better move towards a good solution 👍

**Comment by SW-Vincent** [comment]
> Thanks @SW-Vincent . To answer your questions:
> 
> 1. yes, the idea is to add something like "Only fire this tag if VARIABLE_NAME = XYZ" in the "Configure when the tag should do this" section
> 2. Can you check if your GTM-sync concern is fixed with [Added the option to sync from the GTM dataLayer #879](https://github.com/matomo-org/tag-manager/pull/879) that was recently created?

For the first point i think there is a confusion between perimeters (additionnal conditions for triggers) and trigger groups (or "wait for" triggers) which would both be helpful when it comes to consent (note that each of these 2 features have a dedicated github ticket).

For the second point it indeed fixes my concerns.

> Today I learned that many people & systems pull their consent status straight from the data layer, and users can of course already store that in a custom variable rather than a dedicated named variable. So perhaps having a dedicated consent status variable and a related consent status change trigger aren't as useful.

Maybe dedicated consent status variable would be this useful only if we stick back to the original idea : sending differenciated elements based on consent within a single tag.
